autoload/lsp/internal/completion/documentation.vim

```diff
@@ -65,43 +68,113 @@ endfunction
+            \   'language': &filetype,
+            \   'value': l:completion_item['detail'],
+            \ })
+            let l:documentation = l:detail . "\n\n" . l:documentation
+        endif
     endif
 
-    " TODO: support markdown
-    let l:lines = lsp#utils#_split_by_eol(l:documentation)
-
-    let l:row = float2nr(a:event['row'])
-    let l:col = float2nr(a:event['col'])
-    let l:curpos = screenrow()
+    let l:doc_win = s:get_doc_win()
+    if empty(l:doc_win)
+        call s:close_floating_window(v:true)
+        return
+    endif
 
-    " TODO: reuse floating window/buffer??
-    let s:floating_win = s:FloatingWindow.new()
-    let l:bufnr = s:Buffer.create()
-    call setbufline(l:bufnr, 1, l:lines)
-    call s:floating_win.set_bufnr(l:bufnr)
+    " Update contents.
+    call deletebufline(l:doc_win.get_bufnr(), '^', '$')
+    call setbufline(l:doc_win.get_bufnr(), 1, lsp#utils#_split_by_eol(l:documentation))
+
+    " Calculate layout.
+    let l:size = l:doc_win.get_size({
+    \   'maxwidth': float2nr(&columns * 0.4),
+    \   'maxheight': float2nr(&lines * 0.4),
+    \ })
+    let l:pos = s:compute_position(a:event, l:size)
+    if empty(l:pos)
+        call s:close_floating_window(v:true)
+        return
+    endif
 
-    call s:floating_win.open({
-        \ 'row': 1,
-        \ 'col': 1,
-        \ 'width': 10,
-        \ 'height': 10,
+    if mode()[0] ==# 'i'
+        call l:doc_win.open({
+        \   'row': l:pos[0] + 1,
+        \   'col': l:pos[1] + 1,
+        \   'width': l:size.width,
+        \   'height': l:size.height,
         \ })
+    endif
+    call s:Window.do(l:doc_win.get_winid(), { -> s:Markdown.apply() })
+endfunction
+
+function! s:close_floating_window(force) abort
+    " Ignore `CompleteDone` if it occurred by `complete()` function because the popup menu will re-appear immediately.
+    let l:ctx = {}
+    function! l:ctx.callback(force) abort
+        let l:doc_win = s:get_doc_win()
+        if empty(l:doc_win)
+            return
+        endif
+
+        if !pumvisible() || mode()[0] !=# 'i' || a:force
+            call l:doc_win.close()
+        endif
+    endfunction
+    call timer_start(1, { -> l:ctx.callback(a:force) })
 endfunction
 
-function! s:close_floating_window() abort
-    if exists('s:floating_win')
-        call s:floating_win.close()
+function! s:compute_position(event, size) abort
+    let l:col_if_right = a:event.col + a:event.width + 1 + (a:event.scrollbar ? 1 : 0)
+    let l:col_if_left = a:event.col - a:size.width - 2
+
+    if a:size.width >= (&columns - l:col_if_right)
+        let l:col = l:col_if_left
+    else
+        let l:col = l:col_if_right
+    endif
+
+    if l:col <= 0
+        return []
+    endif
+    if &columns <= l:col + a:size.width
+        return []
     endif
+
+    return [a:event.row, l:col]
+endfunction
+
+function! s:get_doc_win() abort
+    if exists('s:doc_win')
+        return s:doc_win
+    endif
+    if !s:FloatingWindow.is_available()
```

Comment:
check not required since we check it during `_enable`

Reply:
I've removed it.